// @contentful/node-apps-toolkit/lib/utils/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// package.json
var require_package = __commonJS({
  "package.json"(exports2, module2) {
    module2.exports = {
      name: "@contentful/node-apps-toolkit",
      version: "3.13.1",
      description: "A collection of helpers and utilities for creating NodeJS Contentful Apps",
      main: "lib/index.cjs",
      module: "lib/index.mjs",
      types: "lib/index.d.ts",
      files: [
        "lib/**/*"
      ],
      exports: {
        ".": {
          import: "./lib/index.mjs",
          require: "./lib/index.cjs"
        },
        "./keys": {
          import: "./lib/keys/index.mjs",
          require: "./lib/keys/index.cjs"
        },
        "./requests": {
          import: "./lib/requests/index.mjs",
          require: "./lib/requests/index.cjs"
        },
        "./requests/typings": {
          import: "./lib/requests//typingsindex.mjs",
          require: "./lib/requests/typings/index.cjs"
        },
        "./utils": {
          import: "./lib/utils/index.mjs",
          require: "./lib/utils/index.cjs"
        }
      },
      engines: {
        node: ">=18"
      },
      scripts: {
        prepare: "husky install",
        lint: "eslint --ext .ts ./src",
        "lint:fix": "npm run lint -- --fix",
        pretest: "echo ' \u{1F511} Creating valid keypair for testing' && sh test/make-private-keys.sh &> /dev/null",
        "test:unit": "mocha -r dotenv/config -r ts-node/register ./src/**/*.spec.ts",
        "test:integration": "mocha -r dotenv/config -r ts-node/register --timeout 10000 ./test/**/*.test.ts",
        test: "npm run test:unit && npm run test:integration",
        build: "rm -rf lib && tsup",
        postbuild: "cp ./lib/index.d.ts ./lib/index.d.cts",
        "build:docs": "typedoc --options .typedocrc.json src",
        "start:docs": "serve ./docs",
        prepublishOnly: "npm run build && npm run build:docs",
        "semantic-release": "semantic-release",
        format: "prettier --write ./{src,test}/**/*.ts"
      },
      author: "Contentful GmbH",
      license: "MIT",
      dependencies: {
        "@types/debug": "^4.1.5",
        "contentful-management": "^11.6.1",
        debug: "^4.2.0",
        got: "^11.7.0",
        jsonwebtoken: "^9.0.0",
        "lru-cache": "^10.4.3",
        runtypes: "^5.0.1"
      },
      devDependencies: {
        "@semantic-release/changelog": "6.0.3",
        "@semantic-release/git": "10.0.1",
        "@types/jsonwebtoken": "9.0.6",
        "@types/mocha": "10.0.6",
        "@types/node": "22.5.4",
        "@types/sinon": "17.0.3",
        "@typescript-eslint/eslint-plugin": "7.10.0",
        "@typescript-eslint/parser": "7.10.0",
        base64url: "3.0.1",
        dotenv: "16.4.5",
        eslint: "8.57.0",
        "eslint-config-prettier": "9.1.0",
        "eslint-plugin-prettier": "^5.0.0",
        husky: "9.0.11",
        "lint-staged": "15.2.10",
        mocha: "10.4.0",
        prettier: "3.2.5",
        "semantic-release": "23.1.1",
        serve: "14.2.4",
        sinon: "18.0.0",
        "ts-node": "10.9.2",
        tsup: "^8.3.5",
        typedoc: "0.25.13",
        typescript: "5.4.5"
      },
      "lint-staged": {
        "*.ts": [
          "npm run lint:fix",
          "npm run format"
        ]
      },
      repository: {
        url: "github:contentful/node-apps-toolkit",
        type: "git"
      }
    };
  }
});

// src/utils/index.ts
var utils_exports = {};
__export(utils_exports, {
  HttpError: () => import_got.HTTPError,
  createHttpClient: () => createHttpClient,
  createLogger: () => createLogger,
  createValidateStatusCode: () => createValidateStatusCode
});
module.exports = __toCommonJS(utils_exports);

// src/utils/http.ts
var import_got = __toESM(require("got"));
var config = {
  prefixUrl: process.env.BASE_URL || "https://api.contentful.com",
  retry: { limit: 3 }
};
var createHttpClient = (configOverride = {}) => {
  return import_got.default.extend({ ...config, ...configOverride });
};
var createValidateStatusCode = (allowedStatusCodes) => (response) => {
  if (!allowedStatusCodes.includes(response.statusCode)) {
    console.log(response.body);
    throw new import_got.HTTPError(response);
  }
  return response;
};

// src/utils/logger.ts
var path = __toESM(require("path"));
var import_debug = __toESM(require("debug"));
var SRC_PATH = path.join(__dirname, "..");
var { name: APP_NAME } = require_package();
var getNamespaceFromFilename = (filename) => {
  return filename.split(SRC_PATH)[1].slice(0, -3);
};
var createLogger = (opts) => {
  const ns = "namespace" in opts ? opts.namespace : getNamespaceFromFilename(opts.filename);
  return (0, import_debug.default)(APP_NAME).extend(ns);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  HttpError,
  createHttpClient,
  createLogger,
  createValidateStatusCode
});
